server/athenian/api/controllers/features/github/pull_request.py

```diff
@@ -0,0 +1,204 @@
+            return Metric(False, 0, 0, 0)
+        assert self.may_have_negative_values is not None
+        return Metric(True, *mean_confidence_interval(self.samples, self.may_have_negative_values))
+
+    def analyze(self, times: PullRequestTimes) -> Optional[T]:
+        """Do the actual state update."""
+        raise NotImplementedError
+
+
+class PullRequestMedianMetricCalculator(PullRequestMetricCalculator[T]):
+    """Median calculator."""
+
+    def value(self) -> Metric[T]:
+        """Calculate the current metric value."""
+        if not self.samples:
+            return Metric(False, 0, 0, 0)
+        return Metric(True, *median_confidence_interval(self.samples))
+
+    def analyze(self, times: PullRequestTimes) -> Optional[T]:
+        """Do the actual state update."""
+        raise NotImplementedError
+
+
+calculators: Dict[str, Type[PullRequestMetricCalculator]] = {}
+
+
+def register(name: str):
+    """Keep track of the PR metric calculators."""
+    def register_with_name(cls: Type[PullRequestMetricCalculator]):
+        calculators[name] = cls
+        return cls
+
+    return register_with_name
+
+
+class BinnedPullRequestMetricCalculator(Generic[T]):
+    """Batched metrics calculation on sequential time intervals."""
+
+    def __init__(self, calcs: Sequence[PullRequestMetricCalculator[T]],
+                 time_intervals: Sequence[datetime]):
+        """
+        Initialize a new instance of `BinnedPullRequestMetricCalculator`.
+
+        :param calcs: Metric calculators. Their order matches the order of the results in \
+                      `__call__()`.
+        :param time_intervals: Time interval borders. Each interval spans \
+                               `[time_intervals[i], time_intervals[i + 1]]` (the ending is not \
+                               included except for the last interval).
+        """
+        self.calcs = calcs
+        self.time_intervals = time_intervals
+
+    def __call__(self, items: Iterable[PullRequestTimes]) -> List[Tuple[Metric[T]]]:
+        """
+        Calculate the binned metrics.
+
+        For each time interval we collect the list of PRs created then and measure the specified \
+        metrics.
+        """
+        items = sorted(items, key=lambda x: x.created.best)
+        if not items:
+            return []
+        borders = self.time_intervals
+        if items[-1].created.best > borders[-1]:
+            raise ValueError("there are PRs created after time_to")
+        bins = [[] for _ in borders]
+        pos = 0
+        for item in items:
+            while item.created.best > borders[pos + 1]:
+                pos += 1
+            bins[pos].append(item)
+        result = []
+        calcs = self.calcs
+        for bin in bins:
+            for item in bin:
+                for calc in calcs:
+                    calc(item)
+            result.append(tuple(calc.value() for calc in calcs))
+            for calc in calcs:
+                calc.reset()
```

Comment:
Maybe more readable as follows?
```
@contextmanager
def calculated_bin_metrics(calcs, bin):
    for calc in calcs:
        for item in bin:
            calc(item)

    yield tuple(calc.value() for cal in cals)

    for calc in calcs:
        calc.reset()

...
  
for bin in bins:
    with calculated_bin_metrics(calcs, bin) as bin_metrics:
        result.append(bin_metrics)
```

Reply:
I see the profit of not forgetting to `reset()`, but do you really think that adding a generator and a context manager really makes the code simpler or shorter? I would switch to generators if we did not have to calc all the results at once - this is unfortunately not the case...

Reply:
Not a strong opinion 👍